xarray/core/dataset.py

```diff
@@ -1259,98 +1292,29 @@ def isel_points(self, dim='points', **indexers):
-            # supports vindex https://github.com/numpy/numpy/pull/6075
-            if hasattr(variable.data, 'vindex'):
-                # Special case for dask backed arrays to use vectorised list indexing
-                sel = variable.data.vindex[slices]
-            else:
-                # Otherwise assume backend is numpy array with 'fancy' indexing
-                sel = variable.data[slices]
-            return sel
-
-        def relevant_keys(mapping):
-            return [k for k, v in mapping.items()
-                    if any(d in indexer_dims for d in v.dims)]
-
-        coords = relevant_keys(self.coords)
-        indexers = [(k, np.asarray(v)) for k, v in iteritems(indexers)]
-        indexers_dict = dict(indexers)
-        non_indexed_dims = set(self.dims) - indexer_dims
-        non_indexed_coords = set(self.coords) - set(coords)
-
-        # All the indexers should be iterables
-        # Check that indexers are valid dims, integers, and 1D
-        for k, v in indexers:
-            if k not in self.dims:
-                raise ValueError("dimension %s does not exist" % k)
-            if v.dtype.kind != 'i':
-                raise TypeError('Indexers must be integers')
-            if v.ndim != 1:
-                raise ValueError('Indexers must be 1 dimensional')
-
-        # all the indexers should have the same length
-        lengths = set(len(v) for k, v in indexers)
-        if len(lengths) > 1:
-            raise ValueError('All indexers must be the same length')
-
-        # Existing dimensions are not valid choices for the dim argument
-        if isinstance(dim, basestring):
-            if dim in self.dims:
-                # dim is an invalid string
-                raise ValueError('Existing dimension names are not valid '
-                                 'choices for the dim argument in sel_points')
-
-        elif hasattr(dim, 'dims'):
-            # dim is a DataArray or Coordinate
-            if dim.name in self.dims:
-                # dim already exists
-                raise ValueError('Existing dimensions are not valid choices '
-                                 'for the dim argument in sel_points')
-
-        # Set the new dim_name, and optionally the new dim coordinate
-        # dim is either an array-like or a string
-        if not utils.is_scalar(dim):
-            # dim is array like get name or assign 'points', get as variable
-            dim_name = 'points' if not hasattr(dim, 'name') else dim.name
-            dim_coord = as_variable(dim, name=dim_name)
-        else:
-            # dim is a string
-            dim_name = dim
-            dim_coord = None
-
-        reordered = self.transpose(*(list(indexer_dims) + list(non_indexed_dims)))
-
-        variables = OrderedDict()
+        from .dataarray import DataArray
+        if isinstance(dim, DataArray):
+            indexers = {k: DataArray(v, dims=[dim.name], coords=dim.coords)
```

Comment:
This is close but I don't think it's 100% equivalent. I think the order of non-indexed dimensions can vary. With `isel_points`, these always go to the end.

Compare:
```
In [4]: x = xr.DataArray(dims=['x', 'y', 'z'], data=[[[1]]])

In [5]: x
Out[5]:
<xarray.DataArray (x: 1, y: 1, z: 1)>
array([[[1]]])
Dimensions without coordinates: x, y, z

In [9]: x.isel(y=xr.Variable('a', [0]), z=xr.Variable('a', [0]))
Out[9]:
<xarray.DataArray (x: 1, a: 1)>
array([[1]])
Dimensions without coordinates: x, a

In [10]: x.isel_points(y=[0], z=[0], dim='a')
Out[10]:
<xarray.DataArray (a: 1, x: 1)>
array([[1]])
Dimensions without coordinates: a, x
```

I would probably stick with the original code, and we delete it later when we finish the deprecation cycle. Or you could reorder dimensions on the result.